File: backend/app/routers/manufacturing.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional, List
from app.database import get_db
from app.models.manufacturing import ManufacturingOrder, ConversionPattern, BatchRule, MOStatus
from app.auth import require_manager, require_admin, get_current_user

router = APIRouter(prefix="/api/manufacturing", tags=["manufacturing"])


# ── Manufacturing Orders ───────────────────────────────────────────────────────

def mo_out(m: ManufacturingOrder) -> dict:
    return {
        "id": m.id, "mo_number": m.mo_number, "customer": m.customer,
        "quantity": m.quantity, "status": m.status,
        "size_id": m.size_id,
        "size_mm": m.size.value_mm if m.size else None,
        "design_id": m.design_id,
        "design_code": m.design.code if m.design else None,
        "uid_count": len(m.uids),
        "notes": m.notes,
        "created_at": m.created_at,
    }


@router.get("/orders")
def list_orders(status: Optional[MOStatus] = None, db: Session = Depends(get_db), _=Depends(get_current_user)):
    q = db.query(ManufacturingOrder)
    if status:
        q = q.filter(ManufacturingOrder.status == status)
    return [mo_out(m) for m in q.order_by(ManufacturingOrder.id.desc()).all()]


class MOCreate(BaseModel):
    mo_number: str
    customer: str
    quantity: int
    size_id: Optional[int] = None
    design_id: Optional[int] = None
    notes: Optional[str] = None


@router.post("/orders", status_code=201)
def create_order(body: MOCreate, db: Session = Depends(get_db), user=Depends(require_manager)):  # manager+ only
    if db.query(ManufacturingOrder).filter(ManufacturingOrder.mo_number == body.mo_number).first():
        raise HTTPException(status_code=400, detail="MO number already exists")
    mo = ManufacturingOrder(**body.model_dump(), created_by_id=user.id)
    db.add(mo)
    db.commit()
    db.refresh(mo)
    return mo_out(mo)


@router.get("/orders/{mo_id}/uids")
def list_mo_uids(mo_id: int, db: Session = Depends(get_db), _=Depends(get_current_user)):
    from app.models.uid import UID
    mo = db.query(ManufacturingOrder).filter(ManufacturingOrder.id == mo_id).first()
    if not mo:
        raise HTTPException(status_code=404, detail="MO not found")
    uids = db.query(UID).filter(UID.mo_id == mo_id).all()
    from app.routers.uid import uid_out
    return [uid_out(u) for u in uids]


@router.patch("/orders/{mo_id}/status")
def update_mo_status(mo_id: int, status: MOStatus, db: Session = Depends(get_db), _=Depends(require_manager)):  # manager+ only
    mo = db.query(ManufacturingOrder).filter(ManufacturingOrder.id == mo_id).first()
    if not mo:
        raise HTTPException(status_code=404, detail="MO not found")
    mo.status = status
    db.commit()
    return mo_out(mo)


# ── Conversion Patterns ────────────────────────────────────────────────────────

def pattern_out(p: ConversionPattern) -> dict:
    return {
        "id": p.id, "name": p.name,
        "input_length_mm": p.input_length_mm,
        "output_lengths_mm": p.output_lengths_mm,
        "kerf_mm": p.kerf_mm,
        "num_cuts": p.num_cuts,
        "scrap_mm": p.scrap_mm,
        "is_active": p.is_active,
    }


@router.get("/patterns")
def list_patterns(db: Session = Depends(get_db), _=Depends(get_current_user)):
    return [pattern_out(p) for p in db.query(ConversionPattern).filter(ConversionPattern.is_active == True).all()]


class PatternCreate(BaseModel):
    name: str
    input_length_mm: int
    output_lengths_mm: List[int]
    kerf_mm: int = 3


@router.post("/patterns", status_code=201)
def create_pattern(body: PatternCreate, db: Session = Depends(get_db), _=Depends(require_admin)):
    scrap = body.input_length_mm - sum(body.output_lengths_mm) - ((len(body.output_lengths_mm) - 1) * body.kerf_mm)
    if scrap < 0:
        raise HTTPException(status_code=400, detail=f"Pattern results in negative scrap ({scrap}mm)")
    p = ConversionPattern(**body.model_dump())
    db.add(p)
    db.commit()
    db.refresh(p)
    return pattern_out(p)


@router.patch("/patterns/{pattern_id}/archive")
def archive_pattern(pattern_id: int, db: Session = Depends(get_db), _=Depends(require_admin)):
    p = db.query(ConversionPattern).filter(ConversionPattern.id == pattern_id).first()
    if not p:
        raise HTTPException(status_code=404, detail="Pattern not found")
    p.is_active = False
    db.commit()
    return {"archived": True}

File: backend/src/routes/manufacturing.js
import { Router } from 'express';
import { query, one } from '../db/pool.js';
import { asyncHandler } from '../middleware/error.js';
import { requireAuth, requireManager, requireAdmin, HttpError } from '../middleware/auth.js';
import { UID_SELECT, serializeUid } from '../utils/serializers.js';

const router = Router();

// ── Manufacturing Orders ───────────────────────────────────────────────────
async function moOut(m) {
  const cnt = await one('SELECT COUNT(*)::int AS n FROM uids WHERE mo_id = $1', [m.id]);
  return {
    id: m.id,
    mo_number: m.mo_number,
    customer: m.customer,
    quantity: m.quantity,
    status: m.status,
    size_id: m.size_id,
    size_mm: m.size_mm,
    design_id: m.design_id,
    design_code: m.design_code,
    uid_count: cnt.n,
    notes: m.notes,
    created_at: m.created_at,
  };
}

const MO_SELECT = `
  SELECT mo.*, sz.value_mm AS size_mm, d.code AS design_code
  FROM manufacturing_orders mo
  LEFT JOIN sizes sz ON sz.id = mo.size_id
  LEFT JOIN designs d ON d.id = mo.design_id
`;

router.get(
  '/orders',
  requireAuth,
  asyncHandler(async (req, res) => {
    const { status } = req.query;
    const rows = status
      ? await query(`${MO_SELECT} WHERE mo.status = $1 ORDER BY mo.id DESC`, [status])
      : await query(`${MO_SELECT} ORDER BY mo.id DESC`);
    res.json(await Promise.all(rows.map(moOut)));
  })
);

router.post(
  '/orders',
  requireManager,
  asyncHandler(async (req, res) => {
    const { mo_number, customer, quantity, size_id = null, design_id = null, notes = null } = req.body || {};
    const dup = await one('SELECT id FROM manufacturing_orders WHERE mo_number = $1', [mo_number]);
    if (dup) throw new HttpError(400, 'MO number already exists');
    const created = await one(
      `INSERT INTO manufacturing_orders (mo_number, customer, quantity, size_id, design_id, notes, created_by_id)
       VALUES ($1,$2,$3,$4,$5,$6,$7) RETURNING id`,
      [mo_number, customer, quantity, size_id, design_id, notes, req.user.id]
    );
    const m = await one(`${MO_SELECT} WHERE mo.id = $1`, [created.id]);
    res.status(201).json(await moOut(m));
  })
);

router.get(
  '/orders/:moId/uids',
  requireAuth,
  asyncHandler(async (req, res) => {
    const id = parseInt(req.params.moId, 10);
    const mo = await one('SELECT id FROM manufacturing_orders WHERE id = $1', [id]);
    if (!mo) throw new HttpError(404, 'MO not found');
    const rows = await query(`${UID_SELECT} WHERE u.mo_id = $1 ORDER BY u.id`, [id]);
    res.json(rows.map(serializeUid));
  })
);

router.patch(
  '/orders/:moId/status',
  requireManager,
  asyncHandler(async (req, res) => {
    const id = parseInt(req.params.moId, 10);
    const status = req.query.status || (req.body && req.body.status);
    const mo = await one('SELECT id FROM manufacturing_orders WHERE id = $1', [id]);
    if (!mo) throw new HttpError(404, 'MO not found');
    await query('UPDATE manufacturing_orders SET status = $1 WHERE id = $2', [status, id]);
    const m = await one(`${MO_SELECT} WHERE mo.id = $1`, [id]);
    res.json(await moOut(m));
  })
);

// ── Conversion Patterns ────────────────────────────────────────────────────
function patternOut(p) {
  const outputs = p.output_lengths_mm || [];
  const numCuts = outputs.length - 1;
  const scrap = p.input_length_mm - outputs.reduce((a, b) => a + b, 0) - numCuts * p.kerf_mm;
  return {
    id: p.id,
    name: p.name,
    input_length_mm: p.input_length_mm,
    output_lengths_mm: outputs,
    kerf_mm: p.kerf_mm,
    num_cuts: numCuts,
    scrap_mm: scrap,
    is_active: p.is_active,
  };
}

router.get(
  '/patterns',
  requireAuth,
  asyncHandler(async (req, res) => {
    const rows = await query('SELECT * FROM conversion_patterns WHERE is_active = TRUE ORDER BY id');
    res.json(rows.map(patternOut));
  })
);

router.post(
  '/patterns',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { name, input_length_mm, output_lengths_mm, kerf_mm = 3 } = req.body || {};
    const scrap = input_length_mm - output_lengths_mm.reduce((a, b) => a + b, 0) - (output_lengths_mm.length - 1) * kerf_mm;
    if (scrap < 0) throw new HttpError(400, `Pattern results in negative scrap (${scrap}mm)`);
    const p = await one(
      'INSERT INTO conversion_patterns (name, input_length_mm, output_lengths_mm, kerf_mm) VALUES ($1,$2,$3,$4) RETURNING *',
      [name, input_length_mm, JSON.stringify(output_lengths_mm), kerf_mm]
    );
    res.status(201).json(patternOut(p));
  })
);

router.patch(
  '/patterns/:patternId/archive',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = parseInt(req.params.patternId, 10);
    const p = await one('SELECT id FROM conversion_patterns WHERE id = $1', [id]);
    if (!p) throw new HttpError(404, 'Pattern not found');
    await query('UPDATE conversion_patterns SET is_active = FALSE WHERE id = $1', [id]);
    res.json({ archived: true });
  })
);

export default router;
